File: easy/125. Valid Palindrome/125. Valid Palindrome.ts
function isPalindrome(s: string): boolean {
    // const fs = s.toLowerCase().split("").filter((item) => /^[A-Za-z0-9]*$/.test(item)).join("")
    let fs = ""

    for (const item of s) {
        if (/^[A-Za-z0-9]*$/.test(item))
            fs += item.toLowerCase()
    }

    const n = Math.floor(fs.length / 2)
    let f = n - 1, l = n

    if (fs.length % 2 != 0) {
        l += 1
    }

    while (f >= 0) {
        if (fs[f] != fs[l])
            return false
        f -= 1
        l += 1
    }

    return true
};


console.log(isPalindrome("A man, a plan, a canal: Panama"))
console.log(isPalindrome("race a car"))
console.log(isPalindrome(" "))
console.log(isPalindrome("ab"))
console.log(isPalindrome('Marge, let\'s "[went]." I await {news} telegram.'))
console.log(isPalindrome("0P"))

export { }

File: easy/125. Valid Palindrome/125. Valid Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        clear_str = "".join(filter(str.isalnum, s)).lower()
        print(clear_str)

        n = len(clear_str)
        first = (n // 2) - 1
        second = n // 2

        if len(clear_str) % 2 != 0:
            second += 1

        while first >= 0:
            if clear_str[first] != clear_str[second]:
                return False
            first -= 1
            second += 1

        return True


solution = Solution()
# print(solution.isPalindrome("A man, a plan, a canal: Panama"))
# print(solution.isPalindrome("race a car"))
# print(solution.isPalindrome(" "))
# print(solution.isPalindrome("ab"))
# print(solution.isPalindrome('Marge, let\'s "[went]." I await {news} telegram.'))
print(solution.isPalindrome("0P"))
